pydantic/main.py

```diff
@@ -1094,16 +1095,31 @@ def _calculate_keys(self, *args: Any, **kwargs: Any) -> Any:
 RootModelRootType = typing.TypeVar('RootModelRootType')
 
 
+class _RootModelSpecialValues(enum.Enum):
+    no_value = enum.auto()
+
+
+_RootModelNoValue = _RootModelSpecialValues.no_value
+
+
 class RootModel(BaseModel, typing.Generic[RootModelRootType]):
     __pydantic_root_model__ = True
+    __pydantic_fields_set__ = {'root'}  # It's fine having a set here as it will never change
+    __pydantic_private__ = None
     __pydantic_extra__ = None
 
     root: RootModelRootType
 
-    def __init__(__pydantic_self__, root: RootModelRootType) -> None:  # type: ignore
+    def __init__(
+        __pydantic_self__, root: RootModelRootType | Literal[_RootModelNoValue] = _RootModelNoValue
+    ) -> None:  # type: ignore
         __tracebackhide__ = True
-        # this is required set setattr works on RootModels, should be moved to rust
-        _object_setattr(__pydantic_self__, '__pydantic_fields_set__', {'root'})
+        if root is _RootModelNoValue:
+            root_field = __pydantic_self__.model_fields['root']
```

Comment:
technically in this case we should have

```py
_object_setattr(__pydantic_self__, '__pydantic_fields_set__', set())
```

To match `BaseModel` logic.

Reply:
Right. Thank you. This is the place to fix it. I had a discussion with @dmontagu regarding this.